File: frontend/src/lib/lw-profile.ts
import { Protocols } from '@/schemas/primitives';

export const IS_LIGHTWEIGHT_PROFILE =
  import.meta.env.VITE_XUI_PROFILE === 'lw' || import.meta.env.VITE_XUI_PROFILE === 'lightweight';

export const LIGHTWEIGHT_PROTOCOL_OPTIONS = [
  { value: Protocols.VLESS, label: Protocols.VLESS },
] as const;

export const LIGHTWEIGHT_NETWORK_OPTIONS = [
  { value: 'ws', label: 'WebSocket' },
  { value: 'xhttp', label: 'XHTTP' },
  { value: 'tcp', label: 'RAW (Reality)' },
] as const;

File: internal/web/service/lw_profile.go
package service

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/mhsanaei/3x-ui/v3/internal/config"
	"github.com/mhsanaei/3x-ui/v3/internal/database/model"
)

// validateLightweightInbound keeps the lw distribution deliberately narrow:
// VLESS over XHTTP/TLS and WebSocket/TLS for CDN deployments, or TCP/Reality
// for direct anti-blocking deployments. Other protocols remain available in
// the normal profile and can be re-enabled later without a schema migration.
func validateLightweightInbound(inbound *model.Inbound) error {
	if !config.IsLightweightProfile() || inbound == nil {
		return nil
	}
	if inbound.Protocol != model.VLESS {
		return fmt.Errorf("lightweight profile only supports VLESS (XHTTP, WebSocket, or Reality)")
	}

	var stream struct {
		Network  string `json:"network"`
		Security string `json:"security"`
	}
	if raw := strings.TrimSpace(inbound.StreamSettings); raw != "" {
		if err := json.Unmarshal([]byte(raw), &stream); err != nil {
			return fmt.Errorf("invalid stream settings: %w", err)
		}
	}

	switch {
	case stream.Network == "xhttp" && stream.Security == "tls":
		return nil
	case stream.Network == "ws" && stream.Security == "tls":
		return nil
	case stream.Network == "tcp" && stream.Security == "reality":
		return nil
	default:
		return fmt.Errorf("lightweight profile supports only VLESS XHTTP/TLS, VLESS WebSocket/TLS, or VLESS TCP/Reality")
	}
}
